File: src/hello_world_topic_py/hello_world_topic_py/listener.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class Listener(Node):
    def __init__(self):
        super().__init__('listener')
        self.subscription = self.create_subscription(String, '/hello_world_topic', self.listener_callback, 10)
        self.subscription # 避免未使用变量的警告

    def listener_callback(self, msg):
        self.get_logger().info(f"I heard: '{msg.data}'")


def main(args=None):
    rclpy.init(args=args)
    listener_node = Listener()
    rclpy.spin(listener_node) # 启动节点的事件循环
    listener_node.destroy_node() # 显式清理并关闭节点（可选）
    rclpy.shutdown() # 关闭ROS2


if __name__ == '__main__':
    main()

File: src/hello_world_topic_cpp/src/listener.cpp
#include "hello_world_topic_cpp/listener.h"

CListener::CListener() : Node("listener_node")
{
    m_subscriber = this->create_subscription<std_msgs::msg::String>("/hello_world_topic", 10, 
        std::bind(&CListener::topic_callback, this, std::placeholders::_1));
}

CListener::~CListener()
{
}

void CListener::topic_callback(const std_msgs::msg::String &msg) const
{
    RCLCPP_INFO(this->get_logger(), "I heard: '%s'", msg.data.c_str());
}

int main(int argc, char **argv) 
{
    rclcpp::init(argc, argv);
    auto listener_node = std::make_shared<CListener>();

    rclcpp::spin(listener_node);
    rclcpp::shutdown();
    return 0;
}